### PR TITLE
Add 'recover_via' parser

### PR DESCRIPTION
**Pull request by jfecher**
Adds a new parser (currently named `recover_via` - I'm open to suggestions) which attempts the first parser and returns its result on success. On error it will try the given recovery parser and if that succeeds it will return that result but with the errors from the first parser. In the case both parsers fail, the second is considered to be a failed recovery strategy and only the errors from the first are returned. This allows you to chain multiple successive `recover_via` calls.

This parser is similar to `Parser::or` except that it returns the errors of the first parser even if the second succeeds.

This parser is also similar to `recover_with(Strategy)` though it accepts any parser as an argument rather than only an existing strategy, and is thus more general and easier to extend for users (at least with the current design of Strategy).

### CODE REVIEW

File: src/recovery.rs [hunk 1]
```diff
@@ -341,3 +341,55 @@ impl<I: Clone, O, A: Parser<I, O, Error = E>, S: Strategy<I, O, E>, E: Error<I>>
...
+        stream: &mut StreamOf<I, E>,
+    ) -> PResult<I, O, E> {
+        let pre_state = stream.save();
+
+        #[allow(deprecated)]
+        let (a_errors, a_out) = debugger.invoke(&self.0, stream);
+
+        if a_errors.is_empty() {
+            if let Ok(a_out) = a_out {
+                return (vec![], Ok(a_out));
```
Comment (zesterer) [inline src/recovery.rs:366]: Perhaps return `a_errors` here to take advantage of any existing allocated vector space?

**Comment by zesterer** [comment]
Thanks for the PR! I like this change overall, and I was planning for something similar to become the basis for error recovery in the `zero-copy` branch. I wonder whether this would be best implemented as a recovery strategy though: something like `.recover_with(parser(...))`? Interested to hear your thoughts.

File: src/recovery.rs (hunk 1 above)
Comment (jfecher) [reply src/recovery.rs]: Good idea, fixed.

**Comment by jfecher** [comment]
I was actually thinking the opposite, that we could get rid of `Strategy` altogether in favor of only having `recover_via` or other parsers implemented in terms of it. `skip_until` for example could be redefined as `recover_via(none_of(just(xs)).repeated())`, though it'd still need an end of input check to fail. Using this approach over strategy simplifies the API for users a bit in my eyes, and lets the future of the crate focus on adding more parsers rather than splitting efforts between parsers and recovery strategies.

**Comment by zesterer** [comment]
Yep, this was my plan for `zero-copy`: `nested_delimiters` would then just become a parser primitive. However, I'd rather avoid splitting work between `master` and `zero-copy` for now (it's almost a complete refactor of the crate). Perhaps I can merge this MR, but it might not end up in a stable release, at least before `zero-copy` is merged. Do you have thoughts on that?

**Comment by jfecher** [comment]
This PR implements #77 where the `recover_with` parser given there is defined here as `recover_via`

**Comment by zesterer** [comment]
Okay. Thanks for the PR though, it's useful to be able to point people at it!

File: src/recovery.rs [hunk 2]
```diff
@@ -341,3 +341,50 @@ impl<I: Clone, O, A: Parser<I, O, Error = E>, S: Strategy<I, O, E>, E: Error<I>>
...
+            return (a_errors, a_out);
+        }
+
+        stream.revert(pre_state);
+
+        #[allow(deprecated)]
+        let (mut b_errors, b_out) = debugger.invoke(&self.1, stream);
+
+        if b_out.is_ok() {
+            a_errors.append(&mut b_errors);
```
Comment (zesterer) [inline src/recovery.rs:374]: I guess the question of whether secondary errors from b should be preserved is an open one, but I'd err on the side of saying that they shouldn't. Not certain though.

**Comment by zesterer** [comment]
This is effectively superseded by #233 now. I'm happy to reopen this if things change, however.

**Pull request closed without merging**